api/v1/providers/backups.py

```diff
@@ -87,63 +89,23 @@ def stream_backup(
-                        self.offset = 0
-
-                    def write(self, s):
-                        self.buffer.write(s)
-                        self.offset += len(s)
-
-                    def tell(self): 
-                        return self.offset
-
-                    def close(self):
-                        self.buffer.close()
-
-                    def pop(self):
-                        s = self.buffer.getvalue()
-                        self.buffer.close()
-
-                        self.buffer = io.BytesIO()
-
-                        return s
-                
-                stream = FileObjStream()
-
-                with tarfile.open(mode="w|gz", fileobj=stream) as tar:
-                    for root, dirs, files in os.walk(base_path, topdown=False):
-                        for rel_path in files:
-                            logging.getLogger("api").info(rel_path)
-                            abs_path = base_path.joinpath(rel_path)
-                            tar_info = tar.gettarinfo(abs_path, arcname=rel_path)
-                            tar.addfile(tar_info)  
-
-                            with open(abs_path, "rb") as file:
-                                logging.getLogger("api").info(f"file size: { tar_info.size }")
-
-                                blocks, remainder = divmod(tar_info.size, tarfile.BLOCKSIZE)
-
-                                for b in range(blocks):
-                                    block = file.read(tarfile.BLOCKSIZE)
-
-                                    if len(block) < tarfile.BLOCKSIZE:
-                                        raise exceptions.resource.Unavailable("unexpected end of data")
-                                    
-                                    tar.fileobj.write(block)
-                                    yield stream.pop()
-
-                                if remainder > 0:
-                                    block = file.read(remainder)
-                                    tar.fileobj.write(block)
-                                    tar.fileobj.write(tarfile.NUL * (tarfile.BLOCKSIZE - remainder))
-                                    yield stream.pop()
-
-                                    blocks += 1
-
-                                tar.offset += blocks * tarfile.BLOCKSIZE
-                yield stream.pop()
+                def demote(uid, gid):
+                    os.setgid(gid)
+                    os.setuid(uid)
+
+                tar = subprocess.Popen(["tar", "-cf", "-", base_path.name], stdout=subprocess.PIPE, preexec_fn=demote(backup.uid, backup.uid), cwd=base_path.parent)
+                pigz = subprocess.Popen(["pigz", "--best", "-c", "-f"], stdin=tar.stdout, stdout=subprocess.PIPE, preexec_fn=demote(backup.uid, backup.uid), cwd=base_path.parent)
+
+                sel = selectors.DefaultSelector()
+                sel.register(pigz.stdout, selectors.EVENT_READ)
+
+                chunk_size = os.cpu_count() * 128000 # Cpu Cores * pigz 128K Blocks
```

Comment:
Could you do some experiments with this value, try 32k on its own (not multiplied by cpu count) I think that is the gzip  output block size

Reply:
After playing around with it I found anything less than 128K seems to perform worse than values above 128K. Larger values tend to perform better. I managed to peak at 180MB/s with an average around 150MB/s with c8ee5e1507791149ca229a9858442fec99749937